File: backend/app/schemas/knowledge.py
"""
个人知识库(RAG) Pydantic Schema
"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class DocAddIn(BaseModel):
    """手动添加一篇知识文档"""
    title: str = Field(min_length=1, max_length=200)
    content: str = Field(min_length=1)


class DocOut(BaseModel):
    """知识文档列表项响应

    R6/R7 修复(2026-06-25):补齐 status/update_time,对齐 KnowledgeDoc ORM。
    """
    id: int
    title: str
    source_type: str
    source_ref: Optional[str] = None
    char_count: int = 0
    chunk_count: int = 0
    create_time: datetime
    # R6/R7 修复:补齐 status/update_time,对齐 KnowledgeDoc ORM
    status: str = "active"
    update_time: Optional[datetime] = None

    model_config = {"from_attributes": True}


class SearchIn(BaseModel):
    """知识库检索(联调/演示用)"""
    query: str = Field(min_length=1)
    top_k: int = Field(default=5, ge=1, le=20)


class SearchHitOut(BaseModel):
    content: str
    score: float
    doc_id: int
    title: str
    source_type: str


class KbStatsOut(BaseModel):
    doc_total: int = 0
    chunk_total: int = 0
    by_source: dict = {}
    remote_embedding: bool = False


class SyncResultOut(BaseModel):
    code: int = 0
    issue: int = 0
    forum: int = 0
    feedback: int = 0
    ticket: int = 0
    total: int = 0


class EmbeddingConfigIn(BaseModel):
    """管理员配置 embedding(api_key 留空表示不修改)"""
    base_url: Optional[str] = Field(default=None, max_length=512)
    model: Optional[str] = Field(default=None, max_length=128)
    api_key: Optional[str] = Field(default=None, max_length=512)
    enabled: Optional[bool] = None


class EmbeddingConfigOut(BaseModel):
    base_url: str = ""
    model: str = ""
    enabled: bool = False
    api_key_set: bool = False

File: frontend/src/api/knowledge.ts
import { get, post, del, put } from './http'
import type { Page } from '@/types/common'

export interface KnowledgeDoc {
  id: number
  title: string
  source_type: string
  source_ref?: string | null
  char_count: number
  chunk_count: number
  create_time: string
}

export interface SearchHit {
  content: string
  score: number
  doc_id: number
  title: string
  source_type: string
}

export interface KbStats {
  doc_total: number
  chunk_total: number
  by_source: Record<string, number>
  remote_embedding: boolean
}

export interface SyncResult {
  code: number
  issue: number
  forum: number
  feedback: number
  ticket: number
  total: number
}

export interface EmbeddingConfig {
  base_url: string
  model: string
  enabled: boolean
  api_key_set: boolean
}

/** 知识库文档列表 */
export function getDocs(params?: Record<string, unknown>): Promise<Page<KnowledgeDoc>> {
  return get<Page<KnowledgeDoc>>('/knowledge/docs', params)
}

/** 添加知识文档 */
export function addDoc(data: { title: string; content: string }): Promise<{ id: number; chunk_count: number }> {
  return post<{ id: number; chunk_count: number }>('/knowledge/docs', data)
}

/** 删除知识文档 */
export function deleteDoc(id: number): Promise<void> {
  return del<void>(`/knowledge/docs/${id}`)
}

/** 检索 */
export function searchKnowledge(data: { query: string; top_k?: number }): Promise<SearchHit[]> {
  return post<SearchHit[]>('/knowledge/search', data)
}

/** 从平台数据同步 */
export function syncKnowledge(): Promise<SyncResult> {
  return post<SyncResult>('/knowledge/sync')
}

/** 统计 */
export function getKbStats(): Promise<KbStats> {
  return get<KbStats>('/knowledge/stats')
}

/** 一键重建全部存量切片向量(个人KB+Agent知识库, 唯一超管) */
export function reembedAll() {
  return post<Record<string, number>>('/knowledge/embedding-config/reembed', {})
}

/** 获取 embedding 配置 (管理员) */
export function getEmbeddingConfig(): Promise<EmbeddingConfig> {
  return get<EmbeddingConfig>('/knowledge/embedding-config')
}

/** 更新 embedding 配置 (管理员) */
export function updateEmbeddingConfig(data: Partial<{
  base_url: string; model: string; api_key: string; enabled: boolean
}>): Promise<EmbeddingConfig> {
  return put<EmbeddingConfig>('/knowledge/embedding-config', data)
}
